frontend-v2/src/app/users/user-form.tsx

```diff
@@ -0,0 +1,347 @@
+import { Input } from '@/components/ui/input'
+import { Label } from '@/components/ui/label'
+import { Button } from '@/components/ui/button'
+import {
+  Select,
+  SelectContent,
+  SelectItem,
+  SelectTrigger,
+  SelectValue,
+} from '@/components/ui/select'
+import { API_PATH, apiClient, Role, User, UserWithoutId } from '@/lib/api'
+import { Loader2 } from 'lucide-react'
+
+const roleOptions: Role[] = ['admin', 'organizer', 'attendance', 'non-sfbay']
+
+const userFormSchema = z.object({
+  name: z.string().min(1, 'Name is required'),
+  email: z.string().email('Enter a valid email'),
+  chapterId: z
+    .number()
+    .int()
+    .positive('Select a chapter')
+    .nullable()
+    .refine((val): val is number => val !== null, {
+      message: 'Select a chapter',
+    }),
+  disabled: z.boolean(),
+  roles: z.array(Role),
+})
+
+type UserFormValues = z.infer<typeof userFormSchema>
+
+export function UserForm({ userId }: { userId?: number }) {
+  const router = useRouter()
+  const queryClient = useQueryClient()
+
+  const {
+    data: chapters,
+    isLoading: chaptersLoading,
+    isError: chaptersError,
+    error: chaptersErrorObj,
+  } = useQuery({
+    queryKey: [API_PATH.CHAPTER_LIST],
+    queryFn: apiClient.getChapterList,
+  })
+
+  const {
+    data: user,
+    isLoading: userLoading,
+    isError: userError,
+    error: userErrorObj,
+  } = useQuery<User | undefined>({
+    queryKey: [API_PATH.USERS, userId],
+    queryFn: userId ? () => apiClient.getUser(userId) : async () => undefined,
+    enabled: Boolean(userId),
+  })
+
+  const mutation = useMutation({
+    mutationFn: async (payload: UserWithoutId) => {
+      if (userId) {
+        return apiClient.updateUser({ ...payload, id: userId })
+      }
+      return apiClient.createUser(payload)
+    },
+    onSuccess: (savedUser) => {
+      queryClient.invalidateQueries({ queryKey: [API_PATH.USERS] })
+      queryClient.setQueryData([API_PATH.USERS, savedUser.id], savedUser)
+      toast.success(`${savedUser.email} saved`)
+      router.push('/users')
+    },
+    onError: (err: unknown) => {
+      const message =
+        err instanceof Error ? err.message : 'Unable to save user right now'
+      toast.error(message)
+    },
+  })
+
+  const initialValues: UserFormValues = useMemo(
+    () => ({
+      name: user?.name ?? '',
+      email: user?.email ?? '',
+      chapterId: user?.chapter_id ?? null,
+      disabled: user?.disabled ?? false,
+      roles: user?.roles ?? [],
+    }),
+    [user],
+  )
+
+  const form = useForm<UserFormValues>({
+    defaultValues: initialValues,
+    validators: {
+      onSubmit: userFormSchema,
+    },
+    onSubmit: async ({ value }) => {
+      const parsed = userFormSchema.safeParse(value)
+      if (!parsed.success) {
+        toast.error('Please fix the highlighted errors.')
+        return
+      }
+
+      if (parsed.data.chapterId === null) {
+        toast.error('Select a chapter')
+        return
+      }
```

Comment:
feels strange to me that we have to do this here, but i'm used to react hook form. does tanstack form not have a way to ensure your values are valid automatically?